grove/alpha/amplification/amplification.py

```diff
@@ -14,176 +14,80 @@
-        zero_projection = np.array([[1, 0], [0, 0]])
-        one_projection = np.array([[0, 0], [0, 1]])
-
-        control_true = np.kron(one_projection, target_gate)
-        control_false = np.kron(zero_projection, np.eye(2, 2))
-        control_root_true = np.kron(one_projection, sqrtm(target_gate))
-
-        controlled_gate = control_true + control_false
-        controlled_root_gate = control_root_true + control_false
-        assert np.isclose(controlled_gate, np.dot(controlled_root_gate,
-                                                  controlled_root_gate)).all()
-
-        sqrt_name = "SQRT" + target_gate_name
-        adj_sqrt_name = "ADJ" + sqrt_name
-
-        # Initialize program and populate with gate information
-        p = pq.Program()
-
-        if len(controls) == 1:
-            if "C" + target_gate_name not in defined_gates:
-                p.defgate("C" + target_gate_name, controlled_gate)
-                defined_gates.add("C" + target_gate_name)
-            p.inst(("C" + target_gate_name, controls[0], target))
-
-        else:
-            for gate_name, gate in ((sqrt_name, controlled_root_gate),
-                                    (adj_sqrt_name,
-                                     np.conj(controlled_root_gate.T))):
-                if "C" + gate_name not in defined_gates:
-                    p.defgate("C" + gate_name, gate)
-                    defined_gates.add("C" + gate_name)
-            p.inst(("C" + sqrt_name, controls[-1], target))
-            many_toff, new_defined_gates = controlled_program_builder(
-                controls[:-1], controls[-1], 'NOT', np.array([[0, 1], [1, 0]]),
-                set(defined_gates))
-            p += many_toff
-            defined_gates.union(new_defined_gates)
-
-            p.inst(("C" + adj_sqrt_name, controls[-1], target))
-
-            # Don't redefine all of the gates.
-            many_toff.defined_gates = []
-            p += many_toff
-            many_root_toff, new_defined_gates = controlled_program_builder(
-                controls[:-1], target, sqrt_name, sqrtm(target_gate),
-                set(defined_gates))
-            p += many_root_toff
-            defined_gates.union(new_defined_gates)
-
-        return p, defined_gates
-
-    p = controlled_program_builder(controls, target, gate_name, u)[0]
-    return p
+    for _ in range(num_iter):
+        prog += oracle + algorithm.dagger() + diffusion_operator(qubits) + algorithm
+    return prog
 
 
 def diffusion_operator(qubits):
-    """Constructs the (Grover) diffusion operator on qubits,
-    assuming they are ordered from
-    most significant qubit to least significant qubit.
+    """Constructs the diffusion operator in the Hadamard basis on qubits,
+    assuming they are ordered from most significant qubit to least significant qubit.
 
-    The diffusion operator is the diagonal operator given
-    by (1, -1, -1, ..., -1).
+    This is the diagonal operator given by (1, -1, -1, ..., -1).
```

Comment:
Why doesn't this clarify anything? I'm telling readers what operator this is producing. I can expand, but I don't see why you would want to remove the comment.

Reply:
It's neither helping me to understand what this method is doing, not does it help me to understand the physics and math behind it. It merely tells me the output and leaves me confused why it's not returning a diagonal matrix or anything else. If we just state it's returning the diffusion operator it's fine because I don't need to wonder what the exact representation is. We should avoid given half-explanatory comments in doc-strings and push that to proper documentation.

Reply:
It *does* help you understand what it's doing, in that it's telling you exactly what it's producing. It also *does* help you understand the physics, if you have intuition for what a diagonal operator in the Hadamard basis does. (I don't)

I agree that we shouldn't have half-explanatory comments, but I think this is fully explanatory (well, a definition), as long as I add the return type. I would much rather have a one line definition of an operator there for me to see in the docstring than have to navigate to the docs page to read a definition. In addition, I don't think there is currently a docs page, and something with the name `diffusion_operator` might share a name with something else that we don't mean.

I would like to keep it, but I also think I should clean it up because I think the way it is currently worded makes it wrong.

Reply:
The fact that we have this discussion here is proof enough that this doc-string as is is debatable. Either way, please improve it so someone else will not run into the same confusion and misunderstandings I did.